File: app/src/main/java/com/caitlynwiley/pettracker/models/Pet.java
package com.caitlynwiley.pettracker.models;

import java.util.Calendar;
import java.util.Locale;

public class Pet {

    private String species;
    private String breed;
    private String name;
    private double age;
    private String birthday;
    private String gender;
    private String id;

    public Pet() {
    }

    public Pet(String name, String years, String months, String gender, String species) {
        this.name = name;
        setAge(years, months);
        setGender(gender);
        setSpecies(species);
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getSpecies() {
        return species;
    }

    private void setSpecies(String species) {
        this.species = species;
    }

    public String getBreed() {
        return breed;
    }

    public void setBreed(String breed) {
        this.breed = breed;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getAge() {
        return age;
    }

    private void setAge(String y, String m) {
        double years = Integer.parseInt(y);
        double months = Integer.parseInt(m);
        age = years + (months / 12.0);
    }

    public String getBirthday() {
        return birthday;
    }

    public void setBirthday(Calendar c) {
        this.birthday = String.format(Locale.US, "%2d/%2d/%4d", c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH), c.get(Calendar.YEAR));
    }

    public String getGender() {
        return gender;
    }

    private void setGender(String gender) {
        this.gender = gender;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Pet) {
            Pet other = (Pet) obj;
            return other.getId().equals(this.id);
        } else {
            return false;
        }
    }
}

File: app/src/main/java/com/caitlynwiley/pettracker/models/Pet.kt
package com.caitlynwiley.pettracker.models

import com.caitlynwiley.pettracker.view.screens.PetType
import com.google.gson.annotations.SerializedName

class Pet {
    @SerializedName("species")
    var species: PetType = PetType.OTHER

    @SerializedName("breed")
    var breed: String = ""

    @SerializedName("name")
    var name: String = ""

    @SerializedName("age")
    var age: String = ""
        private set

    @SerializedName("birthMonth")
    var birthMonth: Int = -1

    @SerializedName("birthYear")
    var birthYear: Int = -1

    @SerializedName("gender")
    var gender: Gender = Gender.UNKNOWN
        private set

    @SerializedName("id")
    var id: String = ""

    @SerializedName("primaryOwnerName")
    var primaryOwnerName: String = ""

    constructor()

    constructor(name: String, birthYear: String, birthMonth: String, gender: Gender, species: PetType, breed: String) {
        this.name = name
        this.birthMonth = birthMonth.toInt()
        this.birthYear = birthYear.toInt()
        calculateAge(birthYear, birthMonth)
        this.gender = gender
        this.species = species
        this.breed = breed
    }

    private fun calculateAge(year: String, month: String) {
        val birthYear = year.toInt()
        val birthMonth = month.toInt()
        // TODO: make this not hard-coded
        val currentYear = 2021
        val currentMonth = 7
        var ageMonths = currentMonth - birthMonth
        var ageYears = currentYear - birthYear
        if (ageMonths < 0) {
            ageYears--
            ageMonths += 12
        }
        age = "$ageYears year${if (ageYears > 1) "s" else ""}" +
                when (ageMonths) {
                    0 -> ""
                    1 -> " $ageMonths month"
                    else -> " $ageMonths months"
                }
    }

    override fun equals(other: Any?): Boolean {
        return if (other is Pet) {
            other.id == id
        } else {
            false
        }
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    enum class Gender {
        @SerializedName("male") MALE,
        @SerializedName("female") FEMALE,
        @SerializedName("unknown") UNKNOWN
    }
}